_posts/2019-11-07-tfp-cran/tfp-cran.Rmd

```diff
@@ -0,0 +1,252 @@
+knitr::include_graphics("images/g_aleatoric_relu_8.png")
+```
+
+
+Please see [Adding uncertainty estimates to Keras models with tfprobability](https://blogs.rstudio.com/tensorflow/posts/2019-06-05-uncertainty-estimates-tfprobability/) for a detailed explanation. 
+
+Using a normal distribution layer as the output, we can capture irreducible variability in the data, also known as _aleatoric uncertainty_. A different type of probabilistic layer allows to model what is called _epistemic uncertainty_.
+
+### Putting distributions over network weights
+
+Using _variational layers_, we can make neural networks probabilistic.  A simple example could look like so:
+
+```{r}
+model <- keras_model_sequential() %>%
+  layer_dense_variational(
+    units = 1,
+    make_posterior_fn = posterior_mean_field,
+    make_prior_fn = prior_trainable,
+    kl_weight = 1 / n
+  ) %>%
+  layer_distribution_lambda(function(x)
+    tfd_normal(loc = x, scale = 1))
+```
+
+This defines a network with a single dense layer, containing a single neuron, that has a prior distribution put over its weights. The network will be trained to minimize the KL divergence between that prior and an approximate posterior weight distribution, as well as maximize the probability of the data under the posterior weights. (For details, please again see the aforementioned post.)
+
+As a consequence of this setup, each test run will now yield different predictions. For the above simulated data, we might get an ensemble of predictions, like so:
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body", fig.cap = "Epistemic uncertainty on simulated data (from https://blogs.rstudio.com/tensorflow/posts/2019-06-05-uncertainty-estimates-tfprobability/) "}
+knitr::include_graphics("images/g_epistemic_linear_kl150.png")
+```
+
+
+Variational layers for non-dense layers exist, and we'll see an example next week. Now let's move on to the next type of use cases, from big data to small data, in a way.
+
+## Use case 2: Fitting Bayesian models with Monte Carlo methods
+
+In sciences where data don't abound, Markov Chain Monte Carlo (MCMC) methods are common. We've shown some examples how to this with TFP ([Tadpoles on TensorFlow: Hierarchical partial pooling with tfprobability](https://blogs.rstudio.com/tensorflow/posts/2019-05-06-tadpoles-on-tensorflow/), [Hierarchical partial pooling, continued: Varying slopes models with TensorFlow Probability](https://blogs.rstudio.com/tensorflow/posts/2019-05-24-varying-slopes/), [Modeling censored data with tfprobability](https://blogs.rstudio.com/tensorflow/posts/2019-07-31-censored-data/), best read in this order), as well as tried to explain, in an accessible way, some of the background ([On leapfrogs, crashing satellites, and going nuts: A very first conceptual introduction to Hamiltonian Monte Carlo](https://blogs.rstudio.com/tensorflow/posts/2019-10-03-intro-to-hmc/)). 
```

Comment:
```suggestion
In sciences where data aren't abound, Markov Chain Monte Carlo (MCMC) methods are common. We've shown some examples how to this with TFP ([Tadpoles on TensorFlow: Hierarchical partial pooling with tfprobability](https://blogs.rstudio.com/tensorflow/posts/2019-05-06-tadpoles-on-tensorflow/), [Hierarchical partial pooling, continued: Varying slopes models with TensorFlow Probability](https://blogs.rstudio.com/tensorflow/posts/2019-05-24-varying-slopes/), [Modeling censored data with tfprobability](https://blogs.rstudio.com/tensorflow/posts/2019-07-31-censored-data/), best read in this order), as well as tried to explain, in an accessible way, some of the background ([On leapfrogs, crashing satellites, and going nuts: A very first conceptual introduction to Hamiltonian Monte Carlo](https://blogs.rstudio.com/tensorflow/posts/2019-10-03-intro-to-hmc/)). 
```

Reply:
I changed don't to aren't in the very first sentence. (Sorry, no clue why it doesn't show up in the diff).

Reply:
oh, i didn't even know that one!